oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java

```
@@ -0,0 +1,244 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Url-sourced and file-sourced external account credentials.
+ *
+ * <p>By default, attempts to exchange the 3PI credential for a GCP access token.
+ */
+public class IdentityPoolCredentials extends ExternalAccountCredentials {
+
+  /**
+   * The IdentityPool credential source. Dictates the retrieval method of the 3PI credential, which
+   * can either be through a metadata server or a local file.
+   */
+  @VisibleForTesting
+  static class IdentityPoolCredentialSource extends CredentialSource {
+
+    enum IdentityPoolCredentialSourceType {
+      FILE,
+      URL
+    }
+
+    private String credentialLocation;
+    private IdentityPoolCredentialSourceType credentialSourceType;
+
+    @Nullable private Map<String, String> headers;
+
+    public IdentityPoolCredentialSource(Map<String, Object> credentialSourceMap) {
+      super(credentialSourceMap);
+
+      if (credentialSourceMap.containsKey("file")) {
+        credentialLocation = (String) credentialSourceMap.get("file");
+        credentialSourceType = IdentityPoolCredentialSourceType.FILE;
+      } else {
+        credentialLocation = (String) credentialSourceMap.get("url");
+        credentialSourceType = IdentityPoolCredentialSourceType.URL;
+      }
```

Should we fail early if this input is invalid and doesn't contain `file` or `url`?

---

Good question, I'm not sure. The input here should be from a file generated by gCloud, and should be valid.